### PR TITLE
Fix generic return type inference

### LINKED ISSUE #31443: Regression of incorrect literal inference in generic function in 3.4
**TypeScript Version:**  3.4.1 ... 3.4.5, 3.5.0-dev.20190517
**Search Terms:** enum is not assignable to type

**Code**

```ts
class M<T>
{
}

function test<V>(m: M<V>, f: () => V)
{
}

enum Enum
{
	A,
	B,
}

class ClassWithConvert<T>
{
	constructor(init: T)
	{
	}
	convert<U>(converter: { to: (v: T) => U; from: (v: U) => T; })
	{
	}
}

const m2 = new M<ClassWithConvert<Enum>>();
test(m2, () => new ClassWithConvert(Enum.A));
```

**Expected behavior:**
no errors

**Actual behavior:**
Error:

```
Type 'ClassWithConvert<Enum.A>' is not assignable to type 'ClassWithConvert<Enum>'.
  Type 'Enum' is not assignable to type 'Enum.A'.
```

[**Playground Link**](https://www.typescriptlang.org/play/#src=class%20M%3CT%3E%0D%0A%7B%0D%0A%7D%0D%0A%0D%0Afunction%20test%3CV%3E(m%3A%20M%3CV%3E%2C%20f%3A%20()%20%3D%3E%20V)%0D%0A%7B%0D%0A%7D%0D%0A%0D%0Aenum%20Enum%0D%0A%7B%0D%0A%09A%2C%0D%0A%09B%2C%0D%0A%7D%0D%0A%0D%0Aclass%20ClassWithConvert%3CT%3E%0D%0A%7B%0D%0A%09constructor(init%3A%20T)%0D%0A%09%7B%0D%0A%09%7D%0D%0A%09convert%3CU%3E(converter%3A%20%7B%20to%3A%20(v%3A%20T)%20%3D%3E%20U%3B%20from%3A%20(v%3A%20U)%20%3D%3E%20T%3B%20%7D)%0D%0A%09%7B%0D%0A%09%7D%0D%0A%7D%0D%0A%0D%0Aconst%20m2%20%3D%20new%20M%3CClassWithConvert%3CEnum%3E%3E()%3B%0D%0Atest(m2%2C%20()%20%3D%3E%20new%20ClassWithConvert(Enum.A))%3B%0D%0A) (**Note:** Option `strictFunctionTypes` must be checked!)

**Related Issues:** 
These issues look similar: #31204 #28102

When I tried to compile with `typescript@next` (3.5.0-dev.20190517) I've got more detailed error description:

```
error TS2322: Type 'ClassWithConvert<Enum.A>' is not assignable to type 'ClassWithConvert<Enum>'.
  Types of property 'convert' are incompatible.
    Type '<U>(converter: { to: (v: Enum.A) => U; from: (v: U) => Enum.A; }) => void' is not assignable to type '<U>(converter: { to: (v: Enum) => U; from: (v: U) => Enum; }) => void'.
      Types of parameters 'converter' and 'converter' are incompatible.
        Type '{ to: (v: Enum) => any; from: (v: any) => Enum; }' is not assignable to type '{ to: (v: Enum.A) => any; from: (v: any) => Enum.A; }'.
          Types of property 'from' are incompatible.
            Type '(v: any) => Enum' is not assignable to type '(v: any) => Enum.A'.
              Type 'Enum' is not assignable to type 'Enum.A'.
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #31443.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot run dt
@typescript-bot test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20864,7 +20858,10 @@ namespace ts {
                     inferTypes(context.inferences, inferenceSourceType, inferenceTargetType, InferencePriority.ReturnType);
                     // Create a type mapper for instantiating generic contextual types using the inferences made
                     // from the return type.
-                    context.returnMapper = getMapperFromContext(cloneInferredPartOfContext(context));
+                    const returnContext = createInferenceContext(signature.typeParameters!, signature, context.flags);
+                    const returnSourceType = instantiateType(contextualType, outerContext && outerContext.returnMapper);
+                    inferTypes(returnContext.inferences, returnSourceType, inferenceTargetType);
```
Comment (weswigham) [inline src/compiler/checker.ts]: The extra invocation of inference here probably warrants a comment as to why it's needed, especially seeing as its removal affects only the one test just added. 😄

**Comment by weswigham** [comment]
@typescript-bot test this again - afaik none of the auth tokens need to be refreshed for months, so maybe it was just a hiccup in the auth service?

**Comment by ahejlsberg** [comment]
@typescript-bot perf test

**Comment by weswigham** [comment]
@typescript-bot test this but now I've refreshed the credentials so if you don't work I'll _know_ something's wrong.

**Comment by weswigham** [comment]
> Monaco - node (v8.9.0, x64)&#x2F;Emit Time was outside of the allowed range (9.32% worse than baseline).‌

@rbuckton this one platform/solution combination is consistently an outlier - do we know what's up with it yet? Its emit times (and only emit) are always mega high variance, comparatively speaking.

**Pull request merged**